[#147] 서버 에러 핸들링 파이프라인 구축

hyuns66 opened the pull request:
## #️⃣연관된 이슈

- #147 

## 📝작업 내용

    아직 정돈되지 않은 초기 버전입니다.
    - API 호출시 발생하는 에러를 데이터계층에서 ApiException으로 패키징한 Result.failure형태의 데이터를  도메인 계층으로 넘깁니다.
    - 도메인계층(UseCase)에서 이를 받아 뷰에서 처리할 ErrorControlFailureException으로 Throw 합니다.
    - 뷰 계층(ViewModel) 에서 ErrorControlFailureException을 catch 해서 처리합니다.
    - 위 과정이 좀 복잡하기 때문에 API 호출 수행하는 UseCase를 템플릿 매서드화 했습니다. -> ApiRequestUseCase



## PR 발행 전 체크 리스트

- [ ] 발행자 확인
- [ ] 프로젝트 설정 확인
- [ ] 라벨 확인
- [ ] 코드 린트 확인

## 스크린샷 (선택)

## 💬리뷰 요구사항(선택)

- 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요
  - TransitsRemoteDataSource 에서 변경점 모두 반영됐으니까 코드플로우 위주로 한번 확인해주세요
  시간내서 설명한번 드리겠습니다

t1nm1ksun committed:
feat#147 : ApiResponse 핸들링 함수 구현

hyuns66 committed:
feat/#147 : 서버에러 뷰계층 전파 로직 템플릿 작성

아직 정돈되지 않은 초기 버전입니다.
- API 호출시 발생하는 에러를 데이터계층에서 ApiException으로 패키징한 Result.failure형태의 데이터를  도메인 계층으로 넘깁니다.
- 도메인계층(UseCase)에서 이를 받아 뷰에서 처리할 ErrorControlFailureException으로 Throw 합니다.
- 뷰 계층(ViewModel) 에서 ErrorControlFailureException을 catch 해서 처리합니다.
- 위 과정이 좀 복잡하기 때문에 API 호출 수행하는 UseCase를 템플릿 매서드화 했습니다. -> ApiRequestUseCase

hyuns66 committed:
chore : linting

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/data/dataremote/model/response/base/ApiErrorResponse.kt:
```diff
@@ -0,0 +1,10 @@
+
+import kotlinx.serialization.SerialName
+
+data class ApiErrorResponse(
+    @SerialName("responseCode") val responseCode: String,
+    @SerialName("timeStamp") val timeStamp: String? = null,
+    @SerialName("path") val path: String? = null,
+    @SerialName("message") val message: String? = null
+)
```
ApiResponse 데이터 클래스랑 구조가 똑같은것 같은데 따로 구현해두신 이유가 있나요>

hyuns66 replied:
path랑 result차이가 있고 사실 큰차이는 아니긴한데 
나중에 확장성이랑 코드명확성을 위해서 분리했어요

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/data/dataremote/model/response/base/ApiException.kt:
```diff
@@ -0,0 +1,39 @@
+        errorCode: String,
+        errorMessage: String
+    ) : ApiException(errorCode, errorMessage)
+}
```
👍

t1nm1ksun commented on app/src/main/java/com/depromeet/team6/data/dataremote/service/LocationsService.kt:
```diff
@@ -25,7 +25,7 @@ interface LocationsService {
         @Query("lat") lat: Double,
         @Query("lon") lon: Double
-    ): ApiResponse<ResponseAddressDto>
+    ): Response<ApiResponse<ResponseAddressDto>>
```
저는 api-error-handling 브랜치 작업할 때 Response<ApiResponse<ResponseDto>> 구조를 두 번 감싸지 않아도 되도록 네이밍은 예시로 BaseResponse라고 할게요! BaseResponse<ResponseDto> 이런식으로 한 번에 처리되게 하려고 했는데 어떻게 생각하시나요??

t1nm1ksun replied:
저 BaseResponse라는 예시는 내부에서 Response타입으로 받고 ApiResponse 로직을 포함하는 녀석을 생각했어요! 추가로 저희가 핸들링하고 싶은 에러로직도 구현하구요!

t1nm1ksun reviewed (approved):
내일 한 번 더 이야기 나누고 실제로 테스트해보면 좋을 것 같습니다!
고생하셨어요!!

hyuns66 committed:
feat/#147 : 에러 핸들링 파이프라인 완성

- BaseViewModel에 다음 로직이 추가되었습니다.
  - API 관련에러 공통 SideEffect로 처리
  - 이외의 에러 발생시 Firebase Crashlytics 로깅
- typealias를 활용한 BaseResponse로 모든 API 인터페이스 응답 감싸주세요
- CommonSideEffect 클래스가 추가되면서 BaseViewModel의 sideEffect 필드타입이 Contract.SideEffect -> UiSideEffect로 승격되었습니다.
  - Contract에 해당하는 SideEffect만 수집되도록 강제하는 로직에서 UiSideEffect를 상속하는 모든 SideEffect가 수집될 수 있도록 수정되었습니다.
  - 이로인해 생길 수 있는 이슈 없는지 한번씩만 체크해주세요. (제가 BaseViewModel 구현자가 아니어서...

hyuns66 committed:
Merge branch 'feat/mypage#113' of https://github.com/depromeet/16th-team6-Android into feat/api-error-handling

hyuns66 committed:
feat/#147 : ApiError 처리를 위한 SideEffect 수정

- CommonSideEffect -> ApiErrorSideEffect로 변경되었습니다.
- Route 토큰 기반으로 이동하는 로직을 Home, Login, Back 임의 이동으로 수정했습니다.
- 추가적인 Navigate SideEffect가 필요하면 그냥 추가해주세요

hyuns66 committed:
chore : linting

hyuns66 committed:
feat#147 : 경로검색 API 오류처리

hyuns66 committed:
Merge branch 'develop' of https://github.com/depromeet/16th-team6-Android into feat/api-error-handling
